src/routes/hydra.ts

```diff
@@ -1,162 +1,209 @@
+  // Means we have just been redirected from Hydra, and are on the login page
+  // We must check the hydra session to see if we can skip login
+
+  // 2. Call Hydra and check the session of this user
+  return hydraAdminEndpoint
+    .getLoginRequest(hydraChallenge)
+    .then(({body}) => {
+      // If hydra was already able to authenticate the user, skip will be true and we do not need to re-authenticate
+      // the user.
+      if (body.skip) {
+        // You can apply logic here, for example update the number of times the user logged in...
+        // Now it's time to grant the login kratosRequest. You could also deny the kratosRequest if something went terribly wrong
+        // (e.g. your arch-enemy logging in...)
+        const acceptLoginRequest = new AcceptLoginRequest()
+        acceptLoginRequest.subject = String(body.subject)
+
+        logger.debug(
+          'Accepting ORY Hydra Login Request because skip is true: ',
+          acceptLoginRequest
+        )
+        return hydraAdminEndpoint
+          .acceptLoginRequest(hydraChallenge, acceptLoginRequest)
+          .then(({body}) => {
+            // All we need to do now is to redirect the user back to hydra!
+            res.redirect(String(body.redirectTo))
+          })
+      } else if (kratosSessionCookie) {
+        // Figuring out the user
+        return (
+          kratosPublicEndpoint
+            // We need to know who the user is for hydra
+            .whoami(req as { headers: { [name: string]: string } })
+            .then(({body}) => {
+              // We need to get the email of the user. We don't want to do that via traits as
+              // they are dynamic. They would be part of the PublicAPI. That's not true
+              // for identity.addresses So let's get it via the AdmninAPI
+              const subject = body.identity.id
+
+              // User is authenticated, accept the LoginRequest and tell Hydra
+              let acceptLoginRequest: AcceptLoginRequest = new AcceptLoginRequest()
+              acceptLoginRequest.subject = subject
+              acceptLoginRequest.context = body
+
+              return hydraAdminEndpoint
+                .acceptLoginRequest(hydraChallenge, acceptLoginRequest)
+                .then(({body}) => {
+                  // All we need to do now is to redirect the user back to hydra!
+                  res.redirect(String(body.redirectTo))
+                })
+            })
+        )
+      }
+
+      logger.error(
+        'Request and hydraChallenge are both set but user is not authenticated. Unknown state!'
+      )
+      return Promise.reject(
+        new Error(
+          "'Request and hydraChallenge are bot set but user is not authenticated. Please try again!'"
+        )
+      )
+    })
+    .catch(next)
 }
 
+const createHydraSession = (requestedScope: string[] = [], context: LoginContext) => {
+  const verifiableAddresses = context.identity.verifiableAddresses || []
```

Comment:
Use the verifiableAddresses to set the email. If either the scope is missing, or the address is not set, do not set any session information.